plugins/crd/crd_queue.go

```diff
@@ -130,39 +149,53 @@ func networkservicechannelWork(plugin *Plugin) {
 		// We define a function here to process a queue item, so that we can
 		// use 'defer' to make sure the message is marked as Done on the queue
 		func(key string) {
-			var obj interface{}
-			var err error
-
 			defer queueNSC.Done(key)
 
 			// Attempt to split the 'key' into namespace and object name
 			namespace, name, err := cache.SplitMetaNamespaceKey(strKey)
 
 			if err != nil {
+				// This is a soft-error, retry up to QUEUE_RETRY_COUNT times
 				plugin.Log.Errorf("Error splitting meta namespace key into parts: %s", err.Error())
-				runtime.HandleError(fmt.Errorf("error splitting meta namespace key into parts: %s", err.Error()))
-				// This is a soft-error, we merely want to forget it and mark it done on the queue
-				queueNSC.Forget(key)
+				if err != nil {
+					if queueNSC.NumRequeues(key) < QUEUE_RETRY_COUNT {
+						plugin.Log.Errorf("Error processing item with key %s, error %v", key, err)
+						queueNSC.AddRateLimited(key)
+					} else {
+						plugin.Log.Errorf("Failed processing item with key %s, error %v, no more retries", key, err)
+						queueNSC.Forget(key)
+					}
+				}
 				return
 			}
 
 			plugin.Log.Infof("Read item '%s/%s' off workqueue. Processing...", namespace, name)
 
-			// Retrieve the latest version in the cache of this alert
-			obj, err = plugin.sharedFactoryNSC.Networkservice().V1().NetworkServiceChannels().Lister().NetworkServiceChannels(namespace).Get(name)
+			// Retrieve the latest version in the cache of this NetworkService. By using
+			// GetByKey() we are able to determine if the item exists, and thus if it was
+			// added or deleted from the queue, and process appropriately
+			item, exists, err := plugin.informerNSC.GetIndexer().GetByKey(strKey)
 
 			if err != nil {
-				plugin.Log.Errorf("Error getting object '%s/%s' from api: %s", namespace, name, err.Error())
-				runtime.HandleError(fmt.Errorf("Error getting object '%s/%s' from api: %s", namespace, name, err.Error()))
-				// This is a hard-error, we'll raise the flag so the plugin catches this and shuts down
-				queueNSC.Forget(key)
-				plugin.queueError <- true
-				return
+				if queueNSC.NumRequeues(key) < QUEUE_RETRY_COUNT {
```

Comment:
Most of this code looks like it is repeated, can it be refactored to a single method?

Reply:
I agree that appears to be the case. Would it be ok to file an issue and refactor it as a separate commit?